### PR TITLE
Feat(pde): Boundary conditions added

### LINKED ISSUE #54: Boundary conditions
We need to figure out how to do boundary conditions.

### PR DESCRIPTION
**Pull request by kuba05**

### CODE REVIEW

File: tests/boundary_test.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    boundary = ConstantDirichletBoundaryCondition(value=value)
+    new_tensor = boundary.apply_to_initial_condition(tensor)
+    np.testing.assert_array_equal(new_tensor.shape, np.array(tensor.shape) + 2)
+
+    mask = np.ones_like(new_tensor, dtype=bool)
+    interior_slices = tuple(slice(1, -1) for _ in range(new_tensor.ndim))
+    mask[interior_slices] = False
+    expected = np.full_like(new_tensor[mask], value)
+    np.testing.assert_equal(new_tensor[mask], expected)
```
Comment (Maneren) [inline tests/boundary_test.py]: `assert_equal` should support broadcasting, so `np.testing.assert_equal(new_tensor[mask], value)` probably works.

File: tests/boundary_test.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    Should increase tensor size by 2 in each axis and adds the given value at boundary.
+    """
+    boundary = ConstantDirichletBoundaryCondition(value=value)
+    new_tensor = boundary.apply_to_initial_condition(tensor)
+    np.testing.assert_array_equal(new_tensor.shape, np.array(tensor.shape) + 2)
+
+    mask = np.ones_like(new_tensor, dtype=bool)
+    interior_slices = tuple(slice(1, -1) for _ in range(new_tensor.ndim))
+    mask[interior_slices] = False
```
Comment (Maneren) [inline tests/boundary_test.py]: This should be a helper

File: tests/boundary_test.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    """
+    boundary = ConstantDirichletBoundaryCondition(value=value)
+    new_tensor = boundary(tensor, time=time, delta_time=delta_time)
+    np.testing.assert_array_equal(new_tensor.shape, tensor.shape)
+
+    mask = np.ones_like(new_tensor, dtype=bool)
+    interior_slices = tuple(slice(1, -1) for _ in range(new_tensor.ndim))
+    mask[interior_slices] = False
+    expected = np.full_like(new_tensor[mask], 0)
+    np.testing.assert_equal(new_tensor[mask], expected)
```
Comment (Maneren) [inline tests/boundary_test.py]: same as above

File: tests/data.py
Comment (Maneren) [inline tests/data.py:1]: Can't the test tensors from `common.py` be used?

File: src/pde_solver/abc/boundary.py
Comment (Maneren) [inline src/pde_solver/abc/boundary.py:1]: Docstring format, please.

File: src/pde_solver/boundary_conditions.py
Comment (Maneren) [inline src/pde_solver/boundary_conditions.py:1]: Also docstring format please.

File: src/pde_solver/boundary_conditions.py [hunk 4]
```diff
@@ -2,26 +2,39 @@
...
+    def apply_to_initial_condition(self, state: NDArray) -> NDArray:
+        padded = np.pad(state, pad_width=1)
+        slices: list[slice | int] = [slice(None)] * padded.ndim
+        for dim in range(padded.ndim):
+            slices[dim] = 0
+            padded[*slices] = self.value
+            slices[dim] = padded.shape[dim] - 1
+            padded[*slices] = self.value
+            slices[dim] = slice(None)
+        return padded
```
Comment (Maneren) [inline src/pde_solver/boundary_conditions.py]:
```suggestion
        return np.pad(state, pad_width=1, constant_values=self.value)
```

File: tests/boundary_test.py
Comment (Maneren) [inline tests/boundary_test.py:1]: Shouldn't we also test, that values inside the tensor stay unchanged?

File: src/pde_solver/boundary_conditions.py [hunk 5]
```diff
@@ -2,26 +2,39 @@
...
+    @override
+    def __call__(self, state_diff: NDArray, time: float, delta_time: float) -> NDArray:
+        slices: list[slice | int] = [slice(None)] * state_diff.ndim
+        for dim in range(state_diff.ndim):
+            slices[dim] = 0
+            state_diff[*slices] = 0
+            slices[dim] = state_diff.shape[dim] - 1
+            state_diff[*slices] = 0
+            slices[dim] = slice(None)
+        return state_diff
```
Comment (Maneren) [inline src/pde_solver/boundary_conditions.py:41]: How does this compare to the masking procedure used in the tests. Is it somehow better or faster?

**Review by Maneren — CHANGES_REQUESTED** [review]

**Comment by Maneren** [comment]
One more. There is an existing interior helper in the operator test file. Consider reusing it an/or maybe re-implementing in the way it's implemented here depending which approach is better.

File: tests/boundary_test.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from pde_solver.boundary_conditions import ConstantDirichletBoundaryCondition
+from pde_solver.pde_types import NDArray
+from tests.common import (
+    TEST_TENSORS,
+    get_boundary_of_a_tensor,
+    get_interior_of_a_tensor,
+)
+
+
+@pytest.mark.parametrize("value", [-2, 0, 124567, 2.4567654345])
```
Comment (Maneren) [inline tests/boundary_test.py]: Could `TEST_SCALARS` be used here?

File: tests/common.py [hunk 7]
```diff
@@ -129,3 +129,51 @@ def assert_default_epsilon(a: EpsilonComparable, b: EpsilonComparable) -> None:
 
     """
     np.testing.assert_allclose(np.asarray(a), np.asarray(b), atol=DEFAULT_EPSILON)
+
+
+def get_interior_of_a_tensor(tensor: NDArray, order: int = 1) -> NDArray:
```
Comment (Maneren) [inline tests/common.py]: Just `tensor_interior` would be a better name.

File: tests/common.py [hunk 8]
```diff
@@ -129,3 +129,51 @@ def assert_default_epsilon(a: EpsilonComparable, b: EpsilonComparable) -> None:
...
+        input tensor.
+
+    """
+    mask = np.zeros_like(tensor, dtype=bool)
+    interior_slices = tuple(slice(order, -order) for _ in range(tensor.ndim))
+    mask[interior_slices] = True
+    return tensor[mask]
+
+
+def get_boundary_of_a_tensor(tensor: NDArray, order: int = 1) -> NDArray:
```
Comment (Maneren) [inline tests/common.py]: Same here.

**Review by Maneren — APPROVED** [review]
